ENH: add `np.printoptions`, a context manager

ev-br opened the pull request:
Just a syntax sugar over the `get_printoptions/set_printoptions` pair:

```
>>> import numpy as np
>>> with np.printoptions(precision=2):
...     print(np.array([2.0]) / 3)
... 
[0.67]
>>> print(np.array([2.0]) / 3)
[0.66666667]
```


Something like this is probably present in everyone's personal toolbox. One can argue this is where this belongs: personal toolboxes and SO answers (there's at least one, likely more).

The argument for having it in numpy is mostly convenience and ease of introducing to newbees: I can see this used fairly early in teaching, especially non-CS students; the implementation is a bit more magical even if is a mere half dozen lines.

EDIT: closes gh-8344

ev-br committed:
ENH: add `np.printoptions`, a context manager

Just a syntax sugar over the get_printoptions/set_printoptions pair.

pv commented:
Are printoptions threadlocal or global? Should they be threadlocal?

eric-wieser commented:
I think this is a duplicate PR, but the original was a little overreaching.

@pv: global, unfortunately

xoviat commented:
Would this work if two threads used the context manager at the same time?

eric-wieser commented:
@xoviat: no, but nor does manually calling the get and set like users might currently do. Worth warning against in the docstrings though.

Also whatever happens, we fall afoul of PEP550 not being implemented, when it comes to coroutines

mhvk commented:
Since I don't think it is possible to set print options locally now, it seems we cannot blame the context manager to be global in scope too. I think this is good to have.

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -273,6 +273,40 @@ def get_printoptions():
 
 
+@contextlib.contextmanager
+def printoptions(*args, **kwds):
```
nit: `**kwargs` is a more common spelling

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -273,6 +273,40 @@ def get_printoptions():
+    --------
+
+    `printoptions` sets print options temporarily, and restores them back
+    at the end of the `with`-block:
```
I think this belongs in the main doc section above this heading. Maybe the one-line summary should be

> A context manager that sets print options for the scope of the `with` block, restoring the old options at the end

charris commented:
i wonder if we should move all the context managers into their own file at some point? Having them scattered about make it a bit trickier to find them.

ev-br committed:
MAINT: address review comments

ev-br committed:
DOC: add a release not for `printoptions`

ev-br commented:
Ugh, missed #3987 completely. I pushed a version which hopefully addresses review comments, the decision is with you guys :-). 

Making array printing options thread-local is separate, and I'd prefer to defer it to a separate PR.

eric-wieser reviewed (approved):
RST formatting might be a little off, but I don't actually know it well enough to be sure. We can always fix it when the release-notes get cleaned up for release

eric-wieser commented on doc/release/1.15.0-notes.rst:
````diff
@@ -15,6 +15,14 @@ New functions
+    >>> with np.printoptions(precision=2):
+    ...     print(np.array([2.0])) / 3
+    [0.67] 
+``` 
````
I think backticks are markdown-only. Two colons after the word "block" might do the right thing.

mhvk replied:
That's correct.

ev-br replied:
Indeed, fixed and force-pushed.

charris commented:
Thanks Evgeni.
